inference_overview.md

```diff
@@ -0,0 +1,450 @@
+`prepareDataForAdRetrieval()` and may be extended to other ad tech UDFs in the
+future. For more details on these functions see [Protected App Signals][11]
+(PAS)
+
+We will support two main ways of running inference:
+
+1. **Inference with embedded models**: Models are encapsulated in ad tech
+  provided JavaScript code
+1. **Inference with externally provided models**: Models are external to the ad
+  tech provided JavaScript code and APIs are provided to access them
+
+### Inference with embedded models
+
+Models can be embedded in the ad tech's implementation of `generateBid()`. The
+ad tech can embed simpler models in their code, either implemented in JavaScript
+or with [WebAssembly][12] (Wasm).
+
+With this method, the model's binary data and all code required to use it for
+inference are bundled in the ad tech's provided code. This method does not
+require fetching the model binary data (such as weights) externally. Instead, it
+relies on all data being bundled along with the code.
+
+This method can support fully custom models. Since the model and code are fully
+controlled by the ad tech, any compatible model format and code for inference
+can be used: code written in JavaScript and code that can be compiled into Wasm.
+
+This method works better for smaller and simpler models.
+
+- Complex models often require large frameworks or libraries, and other
+  initialization, which may incur a higher cost to run inside the JavaScript
+  sandbox.
+- Not all model code can be translated into Wasm or JavaScript in a
+  straightforward way. It may be more difficult to translate complex models and
+  supporting code. There is a tradeoff between model size and memory cost.
+
+### Inference with externally provided models
+
+In this method, we will support loading binary model data from external sources.
+The models will be fetched periodically from a linked source (e.g. [Google Cloud
+Storage buckets][13], [Amazon S3 buckets][14]) and loaded into memory. An API
+will be provided to `generateBid()` to run inference against the loaded models.
+The whole operation of invoking `runInference()`, running predictions using the
+loaded models and returning the values to the caller will take place inside a
+single TEE. Data will not leave the TEE, and will not be observable externally
+either. Outside observers will not be directly able to observe how many times
+inference was invoked, if at all.
+
+We will initially support [Tensorflow][15] and [PyTorch][16] based models.
+
+We will provide the following APIs, which will be usable from ad tech code:
+
+- **Run inference using the external models**: This is the main way to use
+  inference. A batch of inputs can be constructed in ad tech code and passed to
+  the API. Inference will run for each input on the given pre-loaded model and
+  the responses will be returned. The call, for illustrative purposes, may look
+  like the following:
```

Comment:
Is something missing after this line - were you planning to add a diagram?

Reply:
Same with next bullet

Reply:
Yeah, this should be fixed in the later commits (in this same PR).
Apparently the automated conversion tool stripped all the code boxes, but I did put them in manually.